File: proj4a/Bubble-Sort/bubble.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void generate_random_list(int list[], int size, int max_value) {
    for (int i = 0; i < size; i++) {
        list[i] = rand() % (max_value + 1);
    }
}

void bubble_sort(int list[], int size) {
    for (int i = 0; i < size - 1; i++) {
        for (int j = 0; j < size - i - 1; j++) {
            if (list[j] > list[j + 1]) {
                int temp = list[j];
                list[j] = list[j + 1];
                list[j + 1] = temp;
            }
        }
    }
}

int main() {
    srand(time(NULL));

    long int size = 10000;

    int lists[10][size];
    double execution_times[10];

    // Gerar 10 listas aleatórias
    for (int i = 0; i < 10; i++) {
        generate_random_list(lists[i], size, 100);
    }

    // Medir o tempo de execução para ordenar cada lista individualmente
    for (int i = 0; i < 10; i++) {
        clock_t start_time = clock();
        bubble_sort(lists[i], size);
        clock_t end_time = clock();
        execution_times[i] = ((double) (end_time - start_time)) / CLOCKS_PER_SEC;
    }

    // Exibir os tempos de execução
    for (int i = 0; i < 10; i++) {
        printf("Tempo de execução para a lista %d: %f segundos\n", i + 1, execution_times[i]);
    }

    return 0;
}

File: proj4a/Bubble-Sort/bubble.pl
% Predicado para gerar uma lista de comprimento LENGTH com números aleatórios entre 0 e MAX-VALUE
generate_random_list(Length, MaxValue, List) :-
    length(List, Length),
    maplist(random_between(0, MaxValue), List).

% Predicado para gerar NUM-LISTS listas aleatórias, cada uma com comprimento LENGTH e valores até MAX-VALUE
generate_random_lists(NumLists, Length, MaxValue, Lists) :-
    length(Lists, NumLists),
    maplist(generate_random_list(Length, MaxValue), Lists).

% Predicado para ordenar uma lista usando Bubble Sort
bubble_sort(List, Sorted) :-
    bubble_sort(List, [], Sorted).

bubble_sort([], Acc, Acc).
bubble_sort([X], Acc, [X|Acc]).
bubble_sort([X,Y|Rest], Acc, Sorted) :-
    X > Y,
    bubble_sort([X|Rest], [Y|Acc], Sorted).
bubble_sort([X,Y|Rest], Acc, Sorted) :-
    X =< Y,
    bubble_sort([Y|Rest], [X|Acc], Sorted).

% Predicado para ordenar todas as listas e medir o tempo de execução
sort_and_time_lists(Lists) :-
    length(Lists, NumLists),
    sort_and_time_lists(Lists, 1, NumLists).

sort_and_time_lists([], _, _).
sort_and_time_lists([List|Rest], Idx, NumLists) :-
    statistics(walltime, [_ | [_]]),
    bubble_sort(List, _Sorted),
    statistics(walltime, [_ | [ExecutionTime]]),
    format('Tempo de execucao para a lista ~d: ~3f segundos~n', [Idx, ExecutionTime / 1000.0]),
    NextIdx is Idx + 1,
    sort_and_time_lists(Rest, NextIdx, NumLists).

% Predicado principal para gerar listas aleatórias, ordená-las e medir o tempo de execução
main :-
    generate_random_lists(10, 1000, 100, RandomLists), % 10 listas, comprimento 1000, valores até 100
    sort_and_time_lists(RandomLists).

